go/database/mpt/verification.go

```diff
@@ -570,6 +622,72 @@ func verifyHashesStoredWithParents[N any](
 	return nil
 }
 
+// verifyContractCodes runs list of validation checks on Mpt and on the
+// code file stored in the given directory. These checks include:
+// 1) Fatal checks
+// - All CodeHashes within the Mpt are present in the code file
+// - All CodeHashes within the code file are correct
+// 2) Non-fatal checks
+// - There are no extra Code Hashes not referenced by any account
+func verifyContractCodes(directory string, source *verificationNodeSource, observer VerificationObserver) error {
+	observer.Progress(fmt.Sprintf("Checking contract codes ..."))
+
+	codeFile := filepath.Join(directory, "codes.dat")
+	codes, err := readCodes(codeFile)
+	if err != nil {
+		return err
+	}
+
+	found := make(map[common.Hash]bool)
+	check := func(acc *AccountNode) error {
+		codeHash := acc.info.CodeHash
+		// skip accounts that are not contracts
+		if codeHash == emptyCodeHash {
+			return nil
+		}
+		// no need to check the hash correctness more than once
+		if _, exists := found[codeHash]; exists {
+			return nil
+		}
+
+		// check that the code hash is present in the code file
+		byteCode, exists := codes[codeHash]
+		if !exists {
+			return fmt.Errorf("the hash %x is present in the mpt but missing from the code file", codeHash)
+		}
+		// check correctness of the code hash
+		if got, want := common.Keccak256(byteCode), &codeHash; got.Compare(want) != 0 {
+			return fmt.Errorf("unexpected code hash for address, got: %x want: %x", got, want)
+		}
+		// mark a found hash
+		found[codeHash] = true
+
+		return nil
+	}
+
+	if err := source.forAccountNodes(check); err != nil {
+		return err
+	}
+	// check if there are any contracts within the code file that are not referenced by any accounts
+	if len(found) == len(codes) {
+		return nil
+	}
+
+	observer.Progress(fmt.Sprintf("There are %d contracts not referenced by any accounts:", len(codes)-len(found)))
+
+	// find any extra hashes
+	for h, bc := range codes {
+		if _, exists := found[h]; !exists {
+			observer.Progress(fmt.Sprintf("%x\n", h))
+			if got, want := common.Keccak256(bc), &h; got.Compare(want) != 0 {
```

Comment:
not sure if we need to verify codes that are not used

Reply:
I think we should still check the hash matches the contract byte-codes. It could potentionaly tell us more...